rfc-116-store-attachment-data-in-content-items.md

```diff
@@ -0,0 +1,321 @@
+
+This inconistency restricts what we can do with attachments,
+particularly for Whitehall content.  For example, we cannot generate
+comprehensive [schema.org][] metadata for attachments: [see this
+comment][].
+
+Additionally, users of the content API cannot make use of attachments
+without parsing the document body.  This inhibits creative use of our
+content.
+
+We should make metadata for Whitehall attachments available in the
+same way as other publishing applications, but that requires some
+changes into how we represent attachments across the stack.
+
+
+## Proposal
+
+In our content schemas we already have an [`asset_link` type][], which
+is used for attachments in specialist documents and manual sections.
+It's also used in travel advice pages for the page image and a single
+attachment.  This type appears to solve all of our problems:
+[Publishing API can render such attachments with govspeak][], and
+content items can expose this data to enable programmatic use.
+
+However, the metadata Govspeak expects and the `asset_link` type are
+only *informally* the same.  And furthermore, Whitehall has much richer
+attachment metadata than either `asset_link` or Govspeak allow.
+
+Given that `asset_link` is currently used for both images and
+attachments, adding all of the missing metadata to `asset_link` feels
+a mistake: images and attachments have different metadata, so we
+shouldn't conflate the two.
+
+I propose we do this:
+
+1. Split `asset_link` into four new types: `base_attachment_asset`,
+   `attachment_asset`, `specialist_publisher_attachment_asset` and
+   `image_asset`, updating the schemas which use `asset_link`
+   accordingly, with these fields:
+
+   - `base_attachment_asset`:
+     - `url` (required)
+     - `content_type` (required)
+     - `title`
+   - `attachment_asset` (extends `base_attachment_asset`):
+     - no extra fields at this time
+   - `specialist_publisher_attachment_asset` (extends `base_attachment_asset`):
+     - `content_id` (required)
+     - `created_at`
+     - `updated_at`
+   - `image_asset`:
+     - `url` (required)
+     - `content_type` (required)
+
+   **Migration concerns:** This would need to be done in three steps:
+   1. Update manuals-publisher to not set the `created_at` and `updated_at` for attachments.
+   2. Republish all manual sections.
+   3. Update the schemas.
+
+1. Add new optional fields for extra metadata:
+
+   - `base_attachment_asset`:
+     - `accessible`
+     - `alternative_format_contact_email`
+     - `filename`
+     - `file_size`
+     - `locale`
+     - `number_of_pages`
+     - `preview_url`
+   - `attachment_asset`
+     - `id`
+   - `image_asset`
+     - `alt_text`
+     - `caption`
+     - `credit`
+
+   **Migration concerns:** This can be done without breaking
+   compatibility, as all new fields are optional.
+
+1. Add new attachment type for publication attachments:
+
+    `publication_attachment_asset` (extends `attachment_asset`) with fields:
```

Comment:
Minor: it helped me to check Whitehall and realise `Publication` and `Consultation` are both `Publicationesque`, which is presumably the reason for the naming here?

Reply:
Yes, do you think that would be a good name, if it were explained in the RFC?

Reply:
I think `publication_attachment_asset` is fine, and it would be helpful to explain the reasoning briefly.